core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala

```diff
@@ -286,6 +286,7 @@ class DAGScheduler(
       case None =>
         // We are going to register ancestor shuffle dependencies
         getAncestorShuffleDependencies(shuffleDep.rdd).foreach { dep =>
+          assert(!shuffleToMapStage.get(dep.shuffleId).isDefined)
```

Comment:
Yes, I tend to agree that the way what really should be the same Stage gets recreated as a new Stage is undesirable in the DAGScheduler, but the more important point is that, even though it is not optimal, it does produce correct results, and we can't change that for the sake of code that seems more desirable.

That's not to say that some of the efforts that were begun to fix this kind of behavior while still generating correct results shouldn't be driven to completion.